scripts/buildorder.py

```diff
@@ -167,7 +173,8 @@ def recursive_dependencies(self, pkgs_map, dir_root=None):
                 dependency_package = pkgs_map[dependency_name]
                 if dependency_package.dir != dir_root and dependency_package.only_installing and not self.fast_build_mode:
                     continue
-                result += dependency_package.recursive_dependencies(pkgs_map, dir_root)
+                if self.fast_build_mode:
```

Comment:
fast build mode is not a very clear name. Back in the days we built all dependencies from source when building a package (`./build-package.sh libfoo`), and then a faster -i/-I mode was introduced which downloaded dependency debs instead of build them. This became known as "fast build", i.e. `./build-package -i libfoo`

Reply:
> Why did you make this change?

This is part of a change to the dependency compilation scheme. Now, if `fast_build_mode` is not used, `buildorder.py` will return a list of dependencies of only the requested package. All other dependencies will appear every time the `build-package.sh` script is run. This was done so that `build-package.sh` could define circular dependencies in a non-fast_build_mode.  In the previous dependency compilation scheme, it was impossible to determine them.

Reply:
Reopening because I can't tell if this change makes sense without understanding what you're doing at a higher level. Please leave it open until we're all on the same page about why circular dependencies even came into existence.

Reply:
Ah, I'm starting to see why you're doing it this way now.
That actually makes sense.
Though I'm sure ted might have more substantive comments about it still.

So "`fast_build_mode`" is downloading dependencies from the repos instead of compiling them by specifying `-i`/`-I` on the `build-package.sh` invocation.
I can definitely see why doing the cycle detection recursively, as the dependencies recurse, makes sense.